### PR TITLE
Razor compiler emit CSS scope attributes

### LINKED ISSUE #22776: CSS isolation: Razor compiler update to emit CSS scope attributes
This is one of the work items needed for #10170

Given an extra optional "scope ID" string for each `.razor` file being compiled (which is to be supplied by other work external to this issue), we need the Razor compiler to generate an extra call like:

```cs
builder.AddAttribute(seq, ScopeId, null); // adds an empty attribute named ScopeId
```

... for every HTML element being rendered by that component. This applies equally across regular elements, elements inside RenderFragments, and elements that get optimized into pre-stringified HTML markup fragments.

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Implements https://github.com/dotnet/aspnetcore/issues/22776

The actual IR tree update is pretty simple - we just add an extra attribute to every markup element within a component that has an associated CSS scope.

Most of the work here is plumbing the information from MSBuild through various layers within the Razor compiler. @NTaylorMullen @ajaybhargavb are you happy with the way this is being done here? It feels quite special-casey but it doesn't look like there's any other established way to pass through per-document parameters. I did consider passing this information via `RazorCodeGenerationOptions` but that seems only to be used for project-wide settings so doesn't look right to use it for per-document information.

**Blazor team review notes:** This PR means that our MSBuild code in the SDK can choose to supply `CssScope` metadata for any of the `RazorComponent` itemgroup entries. For example:

```xml
    <Content Update="Pages\Counter.razor">
      <!-- TODO: In our SDK, auto-generate this metadata for any component that has a corresponding *.razor.css file on disk -->
      <!-- The exact value doesn't matter, but should be derived from a hash of the OS-independent relative path to the .razor file within the project -->
      <CssScope>b-8d2ksdjwa20</CssScope>
    </Content>
```

... will end up rendering an extra attribute called `b-8d2ksdjwa20` on each of its HTML elements (e.g., `<span b-8d2ksdjwa20>Hey</span>`). This ties in with the CSS selector rewriting implemented in https://github.com/dotnet/aspnetcore/pull/23657 so that CSS rules only apply to elements rendered by this component.

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
I've also added a further optimization into the "minimized attribute" code path, because this becomes an extremely-frequently hit path for components that have scoped CSS. On components with scoped CSS, every single HTML element has at least one minimized attribute.

The FastGrid-from-blank benchmark improves by around 5% due to this optimization. Admittedly that's a pretty extreme case, but it is the sort of thing we're trying to optimize.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs
Comment (captainsafia) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs:319]: What does removing `WriteParameterSeparator` do here functionally?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs
Comment (captainsafia) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs:21]: Can we do `string cssScope = null` here to make this backwards compatible?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs:319]: Writes out `", "`

Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs:319]: Prevents writing out `", "` which means this method no longer assumes there's additional parameters. Aka imagine going from `AddAttribute(-1, def` to `AddAttribute(-1, abc`

Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs:319]: @SteveSandersonMS all this is an orthogonal fix no?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:98]: Is it valuable for the css scope to be settable / an input? Wondering if this is better off acting more like a checksum where it's an agreed upon hashing mechanism of various inputs.

Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:98]: The main thing is that only some, not all, of the components should emit scope attributes. 

If the compiler knew which ones then it could duplicate the hashing logic itself. However since we have to communicate which ones need the scope, I thought we might as well also avoid the need to duplicate logic for computing the scope value.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs
Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs:319]: Technically it could be treated as orthogonal. However the motivating use case for the optimisation is this PR, ie that it will become a much more common code path that now warrants this small bit of streamlining.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs
Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs:21]: We could, but the class is internal so hopefully we don’t need to be concerned with back-compat.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:98]: > The main thing is that only some, not all, of the components should emit scope attributes.

What's the criteria for if a component should emit scope attributes?

Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:98]: Whether or not the developer chooses to declare scoped styles for it, by having a sibling CSS file (eg MyComponent.razor.css).

Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:98]: To be clear, I know we *could* have the Razor compiler look on disk for .css files, but that seems to violate the idea that this sort of thing is controlled and overridable in MSBuild.

Additionally, there’s a use case for the exact format of the scope identifiers (eg length) to be controllable by developers through csproj properties. Passing the identifiers in from MSBuild means we can do this pretty trivially and without needing more parameters and logic inside the Razor compiler.

File: src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs
Comment (javiercn) [inline src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs:406]: Is this a perf optimization?

Comment (javiercn) [reply src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs:406]: I think I answered my own question [here](https://github.com/dotnet/aspnetcore/pull/23703/files#r450629775). You can correct me if I'm wrong.

Comment (SteveSandersonMS) [reply src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs:406]: Yes: https://github.com/dotnet/aspnetcore/pull/23703#issuecomment-654324052

**Review by javiercn — APPROVED** [review]
I'm not an expert, but the changes and the tests look good!

**Comment by SteveSandersonMS** [comment]
Thanks @javiercn. I'd like to get a sign-off from either @NTaylorMullen or @ajaybhargavb before I merge it though!

**Comment by javiercn** [comment]
> Thanks @javiercn. I'd like to get a sign-off from either @NTaylorMullen or @ajaybhargavb before I merge it though!

Sounds reasonable :)

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:98]: > To be clear, I know we _could_ have the Razor compiler look on disk for .css files, but that seems to violate the idea that this sort of thing is controlled and overridable in MSBuild.

Ya I definitely wouldn't want that either. I'm more trying to land on if we're going to need these css files specifically in the future. For instance, we track "import" documents today aka _ViewImports.cshtml, _Imports.razor etc. by declaring them as inputs into the system because they impact the way the top-level Razor file is parsed. If the answer in the future is "we don't need them as an input for any other reason" then this is a totally fine solution. 

So do you envision us needing these files or their filepaths in the future?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs:45]: Why does the css scope need to be on the project item?

Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs:45]: It's a per-document value. `GenerateCommand` knows the value from its MSBuild inputs, but to get this information into the `CodeDocument` it has to go via the `ProjectItem`, because that's how all the per-document values get into the compiler.

If I'm missing a different way this could/should be done, please let me know!

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs
Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:98]: I don't have any reason to envision rzc needing access to the contents of the CSS files. I was hoping to keep it as agnostic as possible to CSS concepts. Just giving it the minimal information needed to emit some extra specific attributes for specified components hopefully avoids coupling it to any more detailed knowledge of CSS.

**Comment by SteveSandersonMS** [comment]
Pinging @NTaylorMullen @ajaybhargavb - do my answers above sound right to you?

**Comment by ajaybhargavb** [comment]
> Pinging @NTaylorMullen @ajaybhargavb - do my answers above sound right to you?

I have the same question as @NTaylorMullen's [comment above](https://github.com/dotnet/aspnetcore/pull/23703#discussion_r451855183). Otherwise consider me signed off as long as @NTaylorMullen is happy. One other comment,
- I didn't see any Razor SDK tests that test the MSBuild portions etc. cc @pranavkm in case he has any thoughts on what to add here.

**Comment by SteveSandersonMS** [comment]
> I have the same question as @NTaylorMullen's comment above. 

I [posted an answer previously](https://github.com/dotnet/aspnetcore/pull/23703#discussion_r452526022). Is there still an open question? If so, could you clarify what?

**Comment by SteveSandersonMS** [comment]
> I didn't see any Razor SDK tests that test the MSBuild portions etc.

The MSBuild targets are a separate work item, not implemented here.

**Comment by ajaybhargavb** [comment]
> I posted an answer previously. Is there still an open question? If so, could you clarify what?

Woah. I swear GitHub hid your response from me 👀. That resolves my concerns

**Comment by mrpmorris** [comment]
Will this be compatible with MyComponent.razor.css files that are generated from scss files during build?

File: src/Razor/Microsoft.NET.Sdk.Razor/src/RazorGenerate.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.NET.Sdk.Razor/src/RazorGenerate.cs:133]: To make the command line arguments a little easier to understand for those looking at the ocmmand line lets re-iterate over Sources after this iteration is over and mention all of the css scopes & their input. That way it's extra clear that the previous command line parameters are 1-to-1. Alternatively we could just write out a cssscope for every item and have the other side of our system understand "empty" scopes mean nothing.

File: src/Razor/Microsoft.NET.Sdk.Razor/src/RazorGenerate.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.NET.Sdk.Razor/src/RazorGenerate.cs:126]: Question not related to this PR but I noticed it while reviewing: @ajaybhargavb / @pranavkm given the `IsNullOrEmpty` check above if someone clears their `DocumentKind` for a Razor item wont that corrupt every other document kind because it will no longer be 1-to-1? I'd imagine if the parser version was >= 3 then we'd always output `Kind` even if it wasn't on the item to ensure the mappings were correct.

File: src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs:135]: 👍

File: src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs:306]:
```suggestion
            var cssScopeAssociations = new Dictionary<string, int>(StringComparer.Ordinal);
```

Also, with the `StringComparer.Ordinal` that doesn't always hold true on windows OS`. Is there a [FilePathComparer](https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.OmniSharpPlugin/FilePathComparer.cs) equivalent available in this MSBuild task?

File: src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs:29]: `cssscopeinput` makes it sound like this is the CSS file. Ditto with the description.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:126]: Not sure if you want to do the work now or not but the "design time" code path provides all of the inputs upfront:
```C#
public override RazorCodeDocument ProcessDesignTime(RazorSourceDocument source, string fileKind, IReadOnlyList<RazorSourceDocument> importSources, IReadOnlyList<TagHelperDescriptor> tagHelpers)
```

If you want the design time experience to understand that those CSS scopes are on the attributes you'll need to update the `ProcessDesignTime` signature and the corresponding consumption sites in aspnetcore-tooling. That being said, there's no reason that can't be done later so i'll leave it up to you.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectItem.cs:45]: Spent some time to look it over more in depth and there's not currently an easy way to change it without doing a more structural change to how everything at the rzc layer works. Makes me sad but what you have is the best fit for the current architecture.

**Review by NTaylorMullen — APPROVED** [review]

File: src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs
Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs:306]: I think it's sufficient to use `StringComparer.OrdinalIgnoreCase` even on Linux in this particular case. That's because we don't support having components whose names differ only by case. Various parts of the system, at compile and runtime (e.g., routing), will not accept that.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs
Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs:126]: There's no design-time equivalent to the CSS scope attributes. It's a no-op at design time.

File: src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs
Comment (SteveSandersonMS) [reply src/Razor/Microsoft.AspNetCore.Razor.Tools/src/GenerateCommand.cs:29]: Good point. Changed `cssscopeinput` to `cssscopedinput` (to clarify that the scoping is something happening to the input, rather than the input being the scope), and updated descriptions. The term `input` here is still applicable because of its correspondence to the other parameter named `input`.

**Pull request merged**
